lib/active_publisher/async/in_memory_adapter/consumer_thread.rb

```diff
@@ -41,27 +40,42 @@ def start_thread
           @thread = ::Thread.new do
             loop do
               # Write "current_messages" so we can requeue should something happen to the consumer.
-              @current_messages << message = queue.pop
+              @current_messages.concat(queue.pop_up_to(20))
 
               begin
-                ::ActivePublisher.publish(message.route, message.payload, message.exchange_name, message.options)
+                # Only open a single connection for each group of messages to an exchange
+                messages_to_retry = []
+                @current_messages.group_by(&:exchange_name).each do |exchange_name, messages|
+                  begin
+                    ::ActivePublisher.publish_all(exchange_name, messages)
+                  ensure
+                    messages_to_retry.concat(messages)
+                  end
+                end
 
                 # Reset
                 @current_messages = []
               rescue *NETWORK_ERRORS
                 # Sleep because connection is down
                 await_network_reconnect
+                @current_messages.concat(messages_to_retry)
 
                 # Requeue and try again.
-                @current_messages.each do |current_message|
-                  queue.push(current_message)
-                end
+                queue.concat(@current_messages) unless @current_messages.empty?
               rescue => unknown_error
+                @current_messages.concat(messages_to_retry)
+                @current_messages.each do |message|
+                  # Degrade to single message publish ... or at least attempt to
+                  begin
+                    ::ActivePublisher.publish(message.route, message.payload, message.exchange_name, message.options)
+                  rescue => error
+                    ::ActivePublisher.configuration.error_handler.call(unknown_error, {:route => message.route, :payload => message.payload, :exchange_name => message.exchange_name, :options => message.options})
```

Comment:
the original code did not add the message back to the queue as the `unknown_error` path was unknown ... maybe this is different in the case of the multiple messages, but I figured the behavior should remain ... thoughts?